src/lib/define-smart-component.js

```diff
@@ -0,0 +1,98 @@
+import { produce } from './immer.js';
+import { defineSmartComponentCore } from './smart-manager.js';
+
+const META = Symbol('META');
+
+/**
+ * @typedef {Element} SmartContainer
+ * @param {object} context
+ */
+
+/**
+ * @param {object} definition
+ * @param {string} definition.selector
+ * @param {object} definition.params
+ * @param {({
+ *   container: SmartContainer,
+ *   component: Element,
+ *   context: object,
+ *   onEvent: (type: string, listener: (detail: object) => void) => void,
+ *   updateComponent: (
+ *     type: string,
+ *     callbackOrObject: ((property: object) => void),
+ *   ) => void,
+ *   signal: AbortSignal,
+ * }) => void} definition.onContextUpdate
+ * @param {AbortSignal} [signal]
```

Comment:
I didn't know this syntax to make a param optional but I think I prefer the `?` syntax:
` @param {AbortSignal?} signal`

Reply:
Yes, I discovered that the `?` syntax is not really documented for JSDoc but we used it here and there.
I'll update it here and in the other places I used the `[optional]` syntax.

https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#handbook-content